#!/usr/bin/env python3
# Copyright 2023 Georg Novotny
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import os

from ament_index_python.packages import get_package_share_directory
from launch import LaunchDescription
from launch.actions import DeclareLaunchArgument, IncludeLaunchDescription, OpaqueFunction
from launch.substitutions import LaunchConfiguration
from launch_ros.actions import Node
from launch.launch_description_sources import PythonLaunchDescriptionSource
from launch.substitutions import Command
import xacro

XACRO_FILE_NAME = "multi_drone.urdf.xacro"
XACRO_FILE_PATH = os.path.join(get_package_share_directory("sjtu_drone_description"),"urdf", XACRO_FILE_NAME)
R_NS = ["drone0","drone1", "drone2"]

def generate_launch_description():
    use_sim_time = LaunchConfiguration("use_sim_time", default="true")
    use_gui = DeclareLaunchArgument("use_gui", default_value="true", choices=["true", "false"], description="Whether to execute gzclient")
    pkg_gazebo_ros = get_package_share_directory('gazebo_ros')
    pkg_env = get_package_share_directory('sjtu_drone_description')
    
    r_1_doc = xacro.process_file(XACRO_FILE_PATH, mappings = {"drone_id":R_NS[1]})
    r_1_desc = r_1_doc.toprettyxml(indent='  ')

    r_2_doc = xacro.process_file(XACRO_FILE_PATH, mappings = {"drone_id":R_NS[2]})
    r_2_desc = r_2_doc.toprettyxml(indent='  ')

    world_file = os.path.join(pkg_env,"worlds", "env.world")

    def launch_gzclient(context, *args, **kwargs):
        if context.launch_configurations.get('use_gui') == 'true':
            return [IncludeLaunchDescription(
                PythonLaunchDescriptionSource(
                    os.path.join(pkg_gazebo_ros, 'launch', 'gzclient.launch.py')
                ),
                launch_arguments={'verbose': 'true'}.items()
            )]
        return []


    return LaunchDescription([
        use_gui,

        Node(
            package='robot_state_publisher',
            executable='robot_state_publisher',
            namespace=R_NS[1],
            parameters=[{'frame_prefix': R_NS[1]+'/','use_sim_time': use_sim_time, 'robot_description': r_1_desc}]
        ),

        Node(
            package='joint_state_publisher',
            executable='joint_state_publisher',
            name=R_NS[1]+"_" +'joint_state_publisher',
            namespace=R_NS[1],
            output='screen',
        ),

        Node(
            package='robot_state_publisher',
            executable='robot_state_publisher',
            namespace=R_NS[2],
            parameters=[{'frame_prefix': R_NS[2]+'/','use_sim_time': use_sim_time, 'robot_description': r_2_desc}]
        ),

        Node(
            package='joint_state_publisher',
            executable='joint_state_publisher',
            name=R_NS[2]+"_" +'joint_state_publisher',
            namespace=R_NS[2],
            output='screen',
        ),


        IncludeLaunchDescription(
            PythonLaunchDescriptionSource(
                os.path.join(pkg_gazebo_ros, 'launch', 'gzserver.launch.py')
            ),
            launch_arguments={'world': world_file,
                              'verbose': "true",
                              'extra_gazebo_args': 'verbose'}.items()
        ),

        OpaqueFunction(function=launch_gzclient),

        Node(
            package="drone_rl",
            executable="drone_manager",
            output="screen"
        ),

        Node(
            package="drone_rl",
            executable="spawn_drone",
            arguments=[r_1_desc, R_NS[1], "0", "0", "0"],
            output="screen"
        ),

        Node(
            package="drone_rl",
            executable="spawn_drone",
            arguments=[r_2_desc, R_NS[2], "2", "0", "0"],
            output="screen"
        ),

        Node(
            package="tf2_ros",
            executable="static_transform_publisher",
            arguments=["0", "0", "0", "0", "0", "0", "world", f"{R_NS[1]}/odom"],
            output="screen"
        ),

        Node(
            package="tf2_ros",
            executable="static_transform_publisher",
            arguments=["2", "0", "0", "0", "0", "0", "world", f"{R_NS[2]}/odom"],
            output="screen"
        ),
    ])